### PR TITLE
Feature/169366221 flux monitor heartbeat

### PR DESCRIPTION
**Pull request by felder-cl**
Allow chainlink operators to configure the Flux Monitor to trigger a Job Run if it has exceeded a specified 'idle time'.  The 'idle time' is the time for which no Job Run has been started.

The behavior is disabled if no idle time is defined.

https://www.pivotaltracker.com/story/show/169366221

### CODE REVIEW

File: core/services/flux_monitor.go [hunk 1]
```diff
@@ -27,6 +28,9 @@ const defaultHTTPTimeout = 5 * time.Second
 // Monitor supports.
 const MinimumPollingInterval = models.Duration(defaultHTTPTimeout)
 
+// MinimumIdleThreshold is the minimum allowed value for the IdleThreshold attribute.
+const MinimumIdleThreshold = models.Duration(defaultHTTPTimeout)
```
Comment (samphilipd) [inline core/services/flux_monitor.go]: Any reason for this choice of default? Seems arbitrary.

Comment (felder-cl) [reply core/services/flux_monitor.go]: True, I've updated it to be greater than or equal to the pollingInterval

File: core/services/flux_monitor_helpers_test.go [hunk 2]
```diff
@@ -29,7 +29,7 @@ func (p *PollingDeviationChecker) ExportedRespondToNewRound(log eth.Log) error {
 }
 
 func (p *PollingDeviationChecker) ExportedPoll() error {
-	return p.poll()
+	return p.poll(p.threshold)
```
Comment (samphilipd) [inline core/services/flux_monitor_helpers_test.go]: Seems weird to call `p.poll()` with the argument `p.threshold` since it's accessible inside the callee function?

File: core/web/job_specs_controller_test.go [hunk 3]
```diff
@@ -333,7 +334,11 @@ func TestJobSpecsController_Create_FluxMonitor_enabled(t *testing.T) {
 	resp, cleanup := client.Post("/v2/specs", bytes.NewBuffer(jsonStr))
 	defer cleanup()
 
-	require.Equal(t, http.StatusText(http.StatusOK), http.StatusText(resp.StatusCode))
+	body, _ := ioutil.ReadAll(resp.Body)
```
Comment (samphilipd) [inline core/web/job_specs_controller_test.go]: This seems unrelated?

File: core/services/flux_monitor.go [hunk 4]
```diff
@@ -247,18 +248,21 @@ type DeviationChecker interface {
...
+	precision     int32
+	runManager    RunManager
+	currentPrice  decimal.Decimal
+	currentRound  *big.Int
+	fetcher       Fetcher
+	delay         time.Duration
+	cancel        context.CancelFunc
+	newRounds     chan eth.Log
+
+	stopC chan struct{}
```
Comment (rupurt) [inline core/services/flux_monitor.go]: What's the significance of `C` in `stopC`? Is it indicating `Checker` or `Channel`?

File: core/services/flux_monitor_helpers_test.go (hunk 2 above)
Comment (rupurt) [reply core/services/flux_monitor_helpers_test.go]: @samphilipd I think this is parameterized because it's called with 0 in an error condition https://github.com/smartcontractkit/chainlink/pull/2233/files#diff-e2a5bfd429ed26127568dfa7da0a4042R351

File: core/services/flux_monitor_test.go [hunk 5]
```diff
@@ -421,6 +471,7 @@ func TestOutsideDeviation(t *testing.T) {
 		{"inside deviation", decimal.NewFromInt(100), decimal.NewFromInt(101), 2, false},
 		{"equal to deviation", decimal.NewFromInt(100), decimal.NewFromInt(102), 2, true},
 		{"outside deviation", decimal.NewFromInt(100), decimal.NewFromInt(103), 2, true},
+		{"outside deviation zero", decimal.NewFromInt(100), decimal.NewFromInt(0), 2, true},
```
Comment (rupurt) [inline core/services/flux_monitor_test.go]: This is perhaps outside the scope of this PR...

But how do you feel about adding test coverage for when threshold is negative?

File: core/web/job_specs_controller_test.go (hunk 3 above)
Comment (coventry) [reply core/web/job_specs_controller_test.go]: It's an improvement to the failure message. I think it's good.

File: core/services/flux_monitor.go (hunk 4 above)
Comment (felder-cl) [reply core/services/flux_monitor.go]: Its for channel, would you prefer another name or have any suggestions?

File: core/web/job_specs_controller_test.go (hunk 3 above)
Comment (felder-cl) [reply core/web/job_specs_controller_test.go]: It might have been better as its own PR.

Its not completely unrelated though.  This error was probably triggered when updating the code, without enough explanation.

File: core/services/flux_monitor_test.go (hunk 5 above)
Comment (felder-cl) [reply core/services/flux_monitor_test.go]: i feel validation and these constraints are outside the scope for sure.  It probably deserves its own story and we are waiting to hear back re: the audit.

File: core/services/flux_monitor.go [hunk 6]
```diff
@@ -309,18 +316,39 @@ func (p *PollingDeviationChecker) Start(ctx context.Context, client eth.Client)
 	return nil
 }
 
+// stopTimer stops and clears the timer as suggested by the documentation.
+func stopTimer(arg *time.Timer) {
+	if !arg.Stop() && len(arg.C) > 0 {
+		// timer's channel may have backlog
+		// refer to: https://developpaper.com/detailed-explanation-of-the-trap-of-timer-in-golang/
```
[the remark this answers is not in the document]
Comment (felder-cl) [reply core/services/flux_monitor.go]: Authoritative yes, better explained, no.  This was the best article I could find on the subject.

I've updated the comment, but I've assumed the reader has an understanding of the operation of the timer entity.  Therefor I've left out what could go wrong.

File: core/services/flux_monitor.go [hunk 7]
```diff
@@ -329,6 +357,7 @@ func (p *PollingDeviationChecker) consume(ctx context.Context, roundSubscription
 func (p *PollingDeviationChecker) Stop() {
 	if p.cancel != nil {
 		p.cancel()
+		<-p.stopC
```
[the remark this answers is not in the document]
Comment (felder-cl) [reply core/services/flux_monitor.go]: I've renamed it to waitOnStop

File: core/services/flux_monitor_test.go [hunk 8]
```diff
@@ -208,6 +208,56 @@ func TestPollingDeviationChecker_PollHappy(t *testing.T) {
...
+		Return(decimal.NewFromFloat(102), nil).
+		Return(decimal.NewFromFloat(103), nil).
+		Run(func(mock.Arguments) {
+			fetchCalled <- struct{}{}
+		})
+
+	err = deviationChecker.Start(context.Background(), ethClient)
+	require.NoError(t, err)
+
+	<-fetchCalled // Called on start
```
[the remark this answers is not in the document]
Comment (felder-cl) [reply core/services/flux_monitor_test.go]: Thanks for pointing this out

[the remark this answers is not in the document]
Comment (felder-cl) [reply core/services/flux_monitor_test.go]: I've improved the test a bit and fixed this problem in https://github.com/smartcontractkit/chainlink/pull/2233/commits/d955cc6a38306a74db8926df8a25b9455cf98364

File: core/services/flux_monitor.go [hunk 9]
```diff
@@ -309,18 +316,39 @@ func (p *PollingDeviationChecker) Start(ctx context.Context, client eth.Client)
...
+			stopTimer(idleThreshold) // Reset expects stopped or expired timer.
+			idleThreshold.Reset(p.idleThreshold)
+		}
+
 		select {
 		case <-ctx.Done():
+			close(p.stopC)
 			return
 		case err := <-roundSubscription.Err():
 			logger.Error(errors.Wrap(err, "checker lost subscription to NewRound log events"))
```
[the remark this answers is not in the document]
Comment (felder-cl) [reply core/services/flux_monitor.go]: I personally cannot say.  @se3000 should we create a follow up task for it?  Has one been created?

File: core/services/flux_monitor.go [hunk 10]
```diff
@@ -309,18 +316,39 @@ func (p *PollingDeviationChecker) Start(ctx context.Context, client eth.Client)
...
 		case log := <-p.newRounds:
 			logger.ErrorIf(p.respondToNewRound(log), "checker unable to respond to new round")
```
[the remark this answers is not in the document]
Comment (felder-cl) [reply core/services/flux_monitor.go]: Maybe, I'm not entirely convinced.  The thing is with this PR, it seems to be surfacing some items which are not directly related to the change.  I like these are documented and worked on, but I would prefer it not to slow down the current review.

File: core/services/flux_monitor.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (se3000) [reply core/services/flux_monitor.go]: @felder-cl You can if you'd like to. It seems like it's in the scope of this story if it's something you'd like to simplify/refactor. Your call, maybe do it now but as a separate PR? We should definitely put some time into it one way or another. I believe the method is relatively new, and we could probably delete some code on our end if it works.

[the remark this answers is not in the document]
Comment (felder-cl) [reply core/services/flux_monitor.go]: I've opted for its own story.  Given it would probably require its own test/s etc., it could be a significant delay.

https://www.pivotaltracker.com/story/show/171234713

File: core/services/flux_monitor.go (hunk 4 above)
Comment (rupurt) [reply core/services/flux_monitor.go]: I like the change to `waitOnStop`. Captures the intent well.

**Comment by thodges-gh** [comment]
It doesn't seem like this new `"idleThreshold"` is actually working.

**Comment by DavidCurrie1113** [comment]
Its legibility to my finiye mind in comparison to yours is very impressive
in relation to etherscan . Ownet and prospective clients will be able to
understand the code as well . Probability of the effectiveness on
advertising is noticeable as well meta ,is portal
Information bullion.

File: core/web/job_specs_controller_test.go (hunk 3 above)
Comment (se3000) [reply core/web/job_specs_controller_test.go]: Possibly use [cltest.AssertServerResponse](https://github.com/smartcontractkit/chainlink/blob/develop/core/internal/cltest/cltest.go)?

**Comment by felder-cl** [comment]
> It doesn't seem like this new `"idleThreshold"` is actually working.

There was a logic error where the idleThreshold was linked to the last time the price was polled, not updated.  This is fixed in https://github.com/smartcontractkit/chainlink/pull/2233/commits/cdbdca355cff9ecdd42b308567f7ff6e945bd405.

Thanks for spotting this.
